api/clients/v2/verification/router_cert_verifier_address_provider.go

```diff
@@ -0,0 +1,54 @@
+package verification
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	clients "github.com/Layr-Labs/eigenda/api/clients/v2"
+	"github.com/Layr-Labs/eigenda/common"
+	binding "github.com/Layr-Labs/eigenda/contracts/bindings/EigenDACertVerifierRouter"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// RouterAddressProvider is a dynamic provider which fetches cert verifier addresses by making eth_calls
+// against the EigenDACertVerifierRouter contract on the given reference block number.
+type RouterAddressProvider struct {
+	routerBinding      *binding.ContractEigenDACertVerifierRouterCaller
+	blockNumberMonitor *BlockNumberMonitor
+}
+
+// Ensure RouterAddressProvider implements clients.CertVerifierAddressProvider
+var _ clients.CertVerifierAddressProvider = &RouterAddressProvider{}
+
+// BuildRouterAddressProvider creates a new RouterAddressProvider instance
+// that implements the clients.CertVerifierAddressProvider interface
+func BuildRouterAddressProvider(routerAddr gethcommon.Address, ethClient common.EthClient, logger logging.Logger) (*RouterAddressProvider, error) {
+	routerBinding, err := binding.NewContractEigenDACertVerifierRouterCaller(routerAddr, ethClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the BlockNumberMonitor
+	blockNumberMonitor, err := NewBlockNumberMonitor(logger, ethClient, time.Second*3)
+	if err != nil {
+		return nil, fmt.Errorf("create block number monitor: %w", err)
+	}
+
+	return &RouterAddressProvider{
+		routerBinding:      routerBinding,
+		blockNumberMonitor: blockNumberMonitor,
+	}, nil
+}
+
+// GetCertVerifierAddress returns the cert verifier address for the given reference block number
+func (rap *RouterAddressProvider) GetCertVerifierAddress(ctx context.Context, referenceBlockNumber uint64) (gethcommon.Address, error) {
+	// Wait for the local client to reach the reference block number
+	if err := rap.blockNumberMonitor.WaitForBlockNumber(ctx, referenceBlockNumber); err != nil {
```

Comment:
Taking another look after our discussion, the block number monitor probably was just a misguided utility from the start.

I think you should just be able to specify block number as part of the `CallOpts`

Reply:
resolving given overlapping convo: https://github.com/Layr-Labs/eigenda/pull/1532#discussion_r2089418300